org.eclipse.xtext.xbase/src/org/eclipse/xtext/xbase/annotations/validation/UnresolvedFeatureCallTypeAwareMessageProvider.java

```diff
@@ -0,0 +1,156 @@
+import org.eclipse.emf.ecore.EStructuralFeature;
+import org.eclipse.xtext.EcoreUtil2;
+import org.eclipse.xtext.common.types.TypesPackage;
+import org.eclipse.xtext.diagnostics.Diagnostic;
+import org.eclipse.xtext.diagnostics.DiagnosticMessage;
+import org.eclipse.xtext.diagnostics.Severity;
+import org.eclipse.xtext.linking.ILinkingDiagnosticMessageProvider;
+import org.eclipse.xtext.linking.impl.IllegalNodeException;
+import org.eclipse.xtext.linking.impl.LinkingDiagnosticMessageProvider;
+import org.eclipse.xtext.xbase.XAbstractFeatureCall;
+import org.eclipse.xtext.xbase.XFeatureCall;
+import org.eclipse.xtext.xbase.XMemberFeatureCall;
+import org.eclipse.xtext.xbase.XbasePackage;
+import org.eclipse.xtext.xbase.annotations.xAnnotations.XAnnotationsPackage;
+import org.eclipse.xtext.xbase.lib.Extension;
+import org.eclipse.xtext.xbase.lib.IterableExtensions;
+import org.eclipse.xtext.xbase.typesystem.IResolvedTypes;
+import org.eclipse.xtext.xbase.typesystem.internal.FeatureLinkHelper;
+import org.eclipse.xtext.xbase.typesystem.internal.TypeAwareLinkingDiagnosticContext;
+import org.eclipse.xtext.xbase.typesystem.references.LightweightTypeReference;
+import org.eclipse.xtext.xbase.util.FeatureCallAsTypeLiteralHelper;
+
+import com.google.common.collect.Lists;
+import com.google.inject.Inject;
+
+/**
+ * @author Dennis Huebner - Initial contribution and API
+ */
+public class UnresolvedFeatureCallTypeAwareMessageProvider extends LinkingDiagnosticMessageProvider {
+	/**
+	 * A user data entry that indicates a broken feature link which could also
+	 * be a type literal, e.g. 'String::CASE_INSENSITIVE'.
+	 */
+	public static final String TYPE_LITERAL = "key:TypeLiteral";
+
+	public static final String FEATURE_CALL = "key:FeatureCall";
+
+	@Inject
+	private FeatureCallAsTypeLiteralHelper typeLiteralHelper;
+
+	@Inject
+	@Extension
+	private FeatureLinkHelper featureLinkHelper;
+
+	@Override
+	public DiagnosticMessage getUnresolvedProxyMessage(
+			ILinkingDiagnosticMessageProvider.ILinkingDiagnosticContext context) {
+		String linkText = null;
+		try {
+			linkText = context.getLinkText();
+		} catch (IllegalNodeException e) {
+			linkText = e.getNode().getText();
+		}
+		if (linkText == null) {
+			return null;
+		}
+		EObject contextObject = context.getContext();
+		if (isStaticMemberCallTarget(contextObject)) {
+			return new DiagnosticMessage(linkText + " cannot be resolved to a type.", Severity.ERROR,
+					Diagnostic.LINKING_DIAGNOSTIC, UnresolvedFeatureCallTypeAwareMessageProvider.TYPE_LITERAL);
+		}
+		if (contextObject instanceof XAbstractFeatureCall) {
+			if (!((XAbstractFeatureCall) contextObject).isOperation()) {
+				return handleUnresolvedFeatureCall(context, ((XAbstractFeatureCall) contextObject), linkText);
+			}
+		}
+		EClass referenceType = context.getReference().getEReferenceType();
+		String msg = linkText + " cannot be resolved "+ getTypeName(referenceType, context.getReference()) + ".";
```

Comment:
The space after `resolved` wasn't there on purpose since `getTypeName` may return the empty string.